From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 26 Nov 2024 17:15:38 -0500
Subject: [PATCH] Leaf Cache part of canHoldFluid result

Cache the result of part of canHoldFluid logic, since there are many state#is in canAnyHoldFluid method,
it uses map contains to do iteration to check whether a block has a specific block tag key,
which the contains iteration call is very expensive if called everytime

In the test, it can improve ~30% performance in ~1577000 times of canHoldAnyFluid calls (~159ms -> ~111ms)

diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 834e27ef2f7b342b074ff9e1e390e02f3ca1c399..7619061ff78483ea8feb166a0a0ea20f2576006e 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -451,6 +451,8 @@ public abstract class BlockBehaviour implements FeatureElement {
         private VoxelShape[] occlusionShapesByFace;
         private boolean propagatesSkylightDown;
         private int lightBlock;
+        private boolean canHoldAnyFluidInternal; // Leaf - Cache part of canHoldFluid result
+        private boolean canHoldAnyFluidInternalInit; // Leaf - Cache part of canHoldFluid result
 
         // Paper start - rewrite chunk system
         private boolean isConditionallyFullOpaque;
@@ -600,6 +602,8 @@ public abstract class BlockBehaviour implements FeatureElement {
 
             this.propagatesSkylightDown = this.owner.propagatesSkylightDown(this.asState());
             this.lightBlock = this.owner.getLightBlock(this.asState());
+            this.canHoldAnyFluidInternal = false; // Leaf - Cache part of canHoldFluid result
+            this.canHoldAnyFluidInternalInit = false; // Leaf - Cache part of canHoldFluid result
             // Paper start - rewrite chunk system
             this.isConditionallyFullOpaque = this.canOcclude & this.useShapeForLightOcclusion;
             // Paper end - rewrite chunk system
@@ -651,6 +655,18 @@ public abstract class BlockBehaviour implements FeatureElement {
             return block != Blocks.COBWEB && block != Blocks.BAMBOO_SAPLING && this.isSolid();
         }
 
+        // Leaf start - Cache part of canHoldFluid result
+        public boolean canHoldAnyFluidInternal() {
+            // Lazy load cache
+            if (!canHoldAnyFluidInternalInit) {
+                canHoldAnyFluidInternal = net.minecraft.world.level.material.FlowingFluid.canHoldAnyFluid(this.asState());
+                canHoldAnyFluidInternalInit = true;
+            }
+
+            return canHoldAnyFluidInternal;
+        }
+        // Leaf end - Cache part of canHoldFluid result
+
         @Deprecated
         public boolean isSolid() {
             return this.legacySolid;
diff --git a/net/minecraft/world/level/material/FlowingFluid.java b/net/minecraft/world/level/material/FlowingFluid.java
index ace1099a12c762b2e73b71dd3551cf351fedf067..e38fca97993c557d657cc0c3bdb3116f7d734ea4 100644
--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -461,7 +461,7 @@ public abstract class FlowingFluid extends Fluid {
     }
 
     private static boolean canHoldFluid(BlockGetter level, BlockPos pos, BlockState state, Fluid fluid) {
-        return canHoldAnyFluid(state) && canHoldSpecificFluid(level, pos, state, fluid);
+        return /*canHoldAnyFluid(state)*/ state.canHoldAnyFluidInternal() && canHoldSpecificFluid(level, pos, state, fluid); // Leaf - Cache part of canHoldFluid result
     }
 
     private static boolean canHoldSpecificFluid(BlockGetter level, BlockPos pos, BlockState state, Fluid fluid) {
